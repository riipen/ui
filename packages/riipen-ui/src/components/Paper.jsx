import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";
import _JSXStyle from "styled-jsx/style";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

const Paper = props => {
  const {
    children,
    classes,
    component: Component,
    elevation,
    padding,
    rounded,
    variant,
    ...other
  } = props;

  const theme = React.useContext(ThemeContext);

  const className = clsx(
    "root",
    classes,
    rounded ? "rounded" : null,
    variant === "outlined" ? "outlined" : "elevation"
  );

  return (
    <React.Fragment>
      <Component className={className} {...other}>
        {children}
      </Component>
      <style jsx>{`
        .root {
          background-color: ${theme.palette.common.white};
          display: inline-block;
          padding: ${theme.spacing[padding]};
        }

        .elevation {
          box-shadow: ${theme.shadows[elevation]};
        }

        .outlined {
          border: 1px solid ${theme.palette.divider};
        }

        .rounded {
          border-radius: ${theme.shape.borderRadius.md};
        }
      `}</style>
    </React.Fragment>
  );
};

Paper.propTypes = {
  /**
   * The content of the component.
   */
  children: PropTypes.node,

  /**
   * List of additional classes to apply to this component.
   */
  classes: PropTypes.array,

  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component: PropTypes.elementType,

  /**
   * The elevation of the box shadow with respect to the current theme.
   */
  elevation: PropTypes.number,

  /**
   * The amount of padding the paper should have.
   */
  padding: PropTypes.number,

  /**
   * If `true`, rounded corners are enabled.
   */
  rounded: PropTypes.bool,

  /**
   * The variant to use.
   */
  variant: PropTypes.oneOf(["elevation", "outlined"])
};

Paper.defaultProps = {
  classes: [],
  component: "div",
  padding: 2,
  rounded: false,
  variant: "elevation"
};

Paper.displayName = "Paper";

export default withClasses(Paper);
