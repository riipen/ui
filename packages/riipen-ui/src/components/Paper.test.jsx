import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import defaultTheme from "../themes/default";

import Paper from "./Paper";

describe("<Paper>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Paper />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Paper />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const wrapper = mount(<Paper />);

      expect(wrapper.find("Paper").props().component).toEqual("div");
      expect(wrapper.find("Paper").props().rounded).toEqual(false);
      expect(wrapper.find("Paper").props().variant).toEqual("elevation");
    });
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = <div>child</div>;

      const wrapper = mount(<Paper>{child}</Paper>);

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classes = ["newClass"];

      const wrapper = mount(<Paper classes={classes} />);

      expect(
        wrapper
          .find("Paper")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClasses decorator", () => {
        const classes = ["riipen", "riipen-paper"];
  
        const wrapper = mount(<Paper />);
  
        expect(
          wrapper
            .find("Paper")
            .props()
            .classes.sort()
        ).toEqual(classes.sort());
      });
  });

  describe("component prop", () => {
    it("sets custom component", () => {
      const component = "span";

      const wrapper = mount(<Paper component={component} />);

      expect(
        wrapper
          .find("Paper")
          .childAt(0)
          .name()
      ).toEqual(component);
    });

    it("gives an error when given an invalid component", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const component = "abcd";

      mount(<Paper component={component} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("elevation prop", () => {
    it("sets elevation box shadow height for the component", () => {
      const elevation = 3;

      const wrapper = mount(<Paper elevation={elevation} />);

      expect(
        wrapper
          .find("Paper")
          .find("JSXStyle")
          .props().dynamic
      ).toContain(defaultTheme.shadows[elevation]);
    });
  });

  describe("padding prop", () => {
    it("sets padding for the component cell", () => {
      const padding = 3;

      const wrapper = mount(<Paper padding={padding} />);

      expect(
        wrapper
          .find("Paper")
          .find("JSXStyle")
          .props().dynamic
      ).toContain(defaultTheme.spacing[padding]);
    });
  });

  describe("rounded prop", () => {
    it("sets class name as rounded when rounded is true", () => {
      const rounded = true;

      const wrapper = mount(<Paper rounded={rounded} />);

      expect(
        wrapper
          .find("Paper")
          .childAt(0)
          .hasClass("rounded")
      ).toEqual(true);
    });
  });

  describe("variant prop", () => {
    it("sets class name as elevation by default", () => {
        const wrapper = mount(<Paper />);
  
        expect(
          wrapper
            .find("Paper")
            .childAt(0)
            .hasClass("elevation")
        ).toEqual(true);
      });

    it("sets class name as outlined for outlined variant", () => {
      const variant = "outlined";

      const wrapper = mount(<Paper variant={variant}  />);

      expect(
        wrapper
          .find("Paper")
          .childAt(0)
          .hasClass("outlined")
      ).toEqual(true);
    });

    it("gives an error when given an invalid variant", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const variant = "default";

      mount(<Paper variant={variant} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

});
