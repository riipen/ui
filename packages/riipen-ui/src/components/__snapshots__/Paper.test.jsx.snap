// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paper> renders correct snapshot 1`] = `
<withClasses(Paper)>
  <Paper
    classes={
      Array [
        "riipen",
        "riipen-paper",
      ]
    }
    component="div"
    padding={2}
    rounded={false}
    variant="elevation"
  >
    <div
      className="jsx-1659224659 root riipen riipen-paper elevation"
    />
    <JSXStyle
      dynamic={
        Array [
          "#fff",
          undefined,
          undefined,
          "#ddd",
          "4px",
        ]
      }
      id="2575703554"
    />
  </Paper>
</withClasses(Paper)>
`;
